src/libstd/sys/unix/process.rs

```diff
@@ -283,6 +363,24 @@ impl Process {
         if !setup(out_fd, libc::STDOUT_FILENO) { fail(&mut output) }
         if !setup(err_fd, libc::STDERR_FILENO) { fail(&mut output) }
 
+        let close_or_leak = |fd: RawFd| {
+            if cfg.leak_fds_whitelist.contains(&fd) {
+                let flag_orig = libc::fcntl(fd, libc::F_GETFD, 0);
+                if flag_orig < 0 {
+                    return;
+                }
+                let flag_new = flag_orig & !libc::FD_CLOEXEC;
+                if flag_orig != flag_new {
+                    let _ = libc::fcntl(fd, libc::F_SETFD, flag_new);
+                }
+            } else {
+                let _ = libc::close(fd);
+            }
+        };
+        if !cfg.leak_fds {
+            walk_open_fd(close_or_leak);
```

Comment:
This function is called just after the fork, when there is no other thread nor any `open(2)` call, so no more FD are created (except the one from `opendir` which is ignored).

This post-fork code may deadlock for OSX and FreeBSD (as says the `child_after_fork` comment) but `walk_open_fd_proc()`  is only called on Linux. Not sure it's worth using the raw `getdents(2)` here.

We could use `fcntl/FD_CLOEXEC` instead of `close(2)` but it imply two syscalls (one to get the flags and another to update them) which should be slower.